13 make navigation keys configurable

martynaRP opened the pull request:
Navigation keys are configurable now.
First thing I had to do was to remove final keyword from the variable key in Move to be able to assign new value. Then I created methods in Move that take care of assigning new values to keys and providing current configuration. Later I took care of analysing input parameter of main method in Tetris to check for custom configuration. I've decided to dedicate the first value of the input array (args[0]) to pass navigation keys, so we have to run the program in the following way:
java -jar jarFileName.jar "key1 key2 key3"
This way, in the future, we can intorduce another options from command line that can be associated with args[i], where i>0.
In the input, each key has to be represented by a single character, we cannot input less or more characters, and characters must be distinct. If we pass null or empty or blank string, the program will launch with default key configuration. The order of the keys is the same as declared in Move enum, so key1 -> NONE, key2 -> LEFT and key3 -> RIGHT.

TJB4epam commented on src/main/java/com/epam/prejap/tetris/game/Move.java:
```diff
@@ -21,4 +26,38 @@ public static Move of(int key) {
+
+    private static char[] getNavigationKeys() {
+        char[] navigationKeys = new char[values().length];
+        Arrays.stream(values()).forEach(val -> navigationKeys[val.ordinal()] = (char) val.key);
```
Yes, streams are cool.
Yes, so are lambdas.
Yes, simple code is even more so. ;-)

In other words: why so complex?

1. enum in an enum, one is default, the other is configurable and you switch between them
2. static array in an enum, update it's values
3. probably other - also simpler than current implementation - are possible as well :-)

martynaRP replied:
I just wanted to make sure that the keys associated with enum values were updated correctly. But you are right, that creating static array with current configuration makes code simpler. I've uptaded the MoveTest so I'm checking if currentKeys are equal to keys associated with each move.

TJB4epam commented on src/main/java/com/epam/prejap/tetris/game/Move.java:
```diff
@@ -21,4 +26,38 @@ public static Move of(int key) {
+
+    /**
+     * Provides current configuration of navigation keys
+     * @return string representation of moves and associated keys
```
What is wrong with a common `toString()` method?

TJB4epam replied:
And yes, it can be used in a loop on values.

martynaRP replied:
Nothing wrong with toString :) I've added it and used it to generate ckey configuration.

TJB4epam commented on src/test/java/com/epam/prejap/tetris/analyst/CommandLineAnalystTest.java:
```diff
@@ -0,0 +1,116 @@
+
+import static org.testng.Assert.assertEquals;
+
+@Test(groups = "CommandLineAnalyst")
```
It's common for a group name to NOT repeat the class name, but to talk of a feature that this group verifies / checks / tests.

martynaRP replied:
Ok, I've grouped bot test in one group - NavigationKeys

TJB4epam commented on src/test/java/com/epam/prejap/tetris/analyst/CommandLineAnalystTest.java:
```diff
@@ -0,0 +1,116 @@
+        };
+    }
+
+    @Test(dataProvider = "correctInput")
```
Is this test a part of the group "CommandLineAnalyst" or not?

martynaRP replied:
Yes, it is. It checks whether, from the side of a user, keys were updated when provided in a correct manner.

TJB4epam commented on src/test/java/com/epam/prejap/tetris/analyst/CommandLineAnalystTest.java:
```diff
@@ -0,0 +1,116 @@
+public class CommandLineAnalystTest {
+
+    @DataProvider
+    public static Object[] correctInput() {
```
Vague / generic name for your DP.

martynaRP replied:
I've changed the name to more specific.

TJB4epam commented on src/test/java/com/epam/prejap/tetris/analyst/CommandLineAnalystTest.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @DataProvider
+    public static Object[] nullOrBlankArgument() {
```
Specific name. Good.

TJB4epam commented on src/test/java/com/epam/prejap/tetris/analyst/CommandLineAnalystTest.java:
```diff
@@ -0,0 +1,116 @@
+        };
+    }
+
+    //    @Test(dataProvider = "incorrectAmountOfValues")
```
Why we have a commented out code pushed to the repo?

martynaRP replied:
ups... sorry, removed

TJB4epam commented on src/test/java/com/epam/prejap/tetris/analyst/CommandLineAnalystTest.java:
```diff
@@ -0,0 +1,116 @@
+    @Test(
+            dataProvider = "incorrectAmountOfValues",
+            expectedExceptions = IllegalArgumentException.class,
+            expectedExceptionsMessageRegExp = "Incorrect amount of values provided for navigation keys."
```
Nice! We have a message even.

TJB4epam commented on src/test/java/com/epam/prejap/tetris/analyst/CommandLineAnalystTest.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @DataProvider
+    public static Object[] inputWithNonCharacterValues() {
```
I'd expect some numbers here. Like, instead of a 'q' it's really easy to input 1 or 2, they're really close.

martynaRP replied:
Actually, all characters are acceptable as keys, even digits.

TJB4epam commented on src/test/java/com/epam/prejap/tetris/analyst/CommandLineAnalystTest.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @DataProvider
+    public static Object[] inputWithRepeatingKeys() {
```
I ADORE the amount of test scenarios you came up with Martyna. Seriously.

TJB4epam commented on src/test/java/com/epam/prejap/tetris/game/MoveTest.java:
```diff
@@ -0,0 +1,44 @@
+        char[] actualKeys = Move.modifyNavigationKeys(providedKeys);
+
+        //then
+        assertNotEquals(actualKeys, defaultKeys, "Navigation keys should be configurable but are not.");
```
This test confused me. I needed to re-read it 3 times. That's at least partially because I'm quite tired, but an improvement that could remedy that is to make the message more explicit by mentioning that the keys are STILL at default values despite attempting to configure them. Perhaps in a less verbose manner.

martynaRP replied:
I hope I've made the test clear now.

TJB4epam commented on src/test/java/com/epam/prejap/tetris/game/MoveTest.java:
```diff
@@ -0,0 +1,44 @@
+        char[] actualKeys = Move.modifyNavigationKeys(providedKeys);
+
+        //then
+        assertEquals(actualKeys, providedKeys, "Navigation keys should have been modified to provided ones but" +
```
It's Java 14 - multiline Strings are an option. Just saying.

TJB4epam reviewed (changes requested):
Martyna - love the amount of tests, needs rework though. 

Also - remember to bump up the POM version!

martynaRP commented on src/main/java/com/epam/prejap/tetris/Tetris.java:
```diff
@@ -1,5 +1,6 @@
 package com.epam.prejap.tetris;
 
+import com.epam.prejap.tetris.analyst.CommandLineAnalyst;
```
I have to admit I wasn't sure were to place this class. My first thougt was to put it just in testris package as it's used only by Tetris.main(). But then I thougt about progam growing and having more inputs from the command line analysed, so I decided to put it in a separate package to keep the code organized. Would it be better if I changed the class name to NavigationKeysAnalyst so it's associated only with checking one functionality? Or I should remove it from the package and put in tetris package, as for now we don't have and don't know if there will be a need of other analysts?

martynaRP commented on src/main/java/com/epam/prejap/tetris/Tetris.java:
```diff
@@ -43,6 +44,14 @@ public Score play() {
+     * Prepares the environment and launches the game
+     *
+     * @param args  array of strings input from the command line
+     *              args[0] dedicated to configure custom navigation keys
```
I've improved the javadocs.

martynaRP commented on src/main/java/com/epam/prejap/tetris/analyst/CommandLineAnalyst.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    public static char[] checkArgsForNavigationKeys(String arg0) {
+        char[] keys = Move.defaultKeys();
+        if (arg0 != null && !arg0.isBlank()) {
```
I've shortened the method by changing if-check.

martynaRP commented on src/main/java/com/epam/prejap/tetris/analyst/CommandLineAnalyst.java:
```diff
@@ -0,0 +1,49 @@
+            if (keys.length == numberOfNavigationKeys) {
+                if (distinctValuesForNavigationKeys(keys)) {
+                    return keys;
+                } else throw new IllegalArgumentException("Each navigation key should be represented by a " +
```
I've reorganized the code.

martynaRP commented:
@TJB4epam Thank you for all your comments! I hope I've managed to improve my code. I'm ready for another review.

martynaRP commented:
@pkierat-epam @TJB4epam 
I've updated my code regarding recent PR merge. Could you please take a look at it?

pkierat-epam reviewed (commented):
The history of your branch needs to be rewritten before it can be merged to master. If you don't mind (and I'm sure you don't :wink:), I'm going to use it as an example of problems you can run into when working with Git during the next lecture (on Wednesday).

In the meantime, you can still add changes if you wish.

martynaRP commented on src/test/java/com/epam/prejap/tetris/JarFileTest.java:
```diff
@@ -0,0 +1,54 @@
+package com.epam.prejap.tetris;
```
Yes, I got really lost when merging changes done by Szymon, and at the same time I had some local commits done and after merging master into my branch, when I tired to push to my remote branch, I got an error and it looked (with a help of intelliJ) like I have to pull from it first, resolve conflicts, and only then push... So yes, it was exactly like you said. But due to lack of my experience with git I went on with it... Now I fixed it by reseting the local branch to my last commit (before merging) and pushing this hard reset to my remote branch. I know we've talked we should avoid it, but as it would affect only my work, I though it was the easiest way.

martynaRP replied:
And no, I don't mind, you can use my situation as an example :)

martynaRP commented on src/main/java/com/epam/prejap/tetris/game/Move.java:
```diff
@@ -1,18 +1,29 @@
-    NONE(' '),
-    LEFT('h'),
-    RIGHT('l'),
+    NONE(defaultKeys()[0]),
```
@pkierat-epam I've changed the code in enum using your second suggestion. I've also rewritten the history on my branch and bumped up the version. Do you have any other suggestions for me to improve my code before it can be merged?

martynaRP commented:
@pkierat-epam I've changed the code in enum using your second suggestion. I've also rewritten the history on my branch and bumped up the version. Do you have any other suggestions for me to improve my code before it can be merged?

pkierat-epam commented:
Looks good! Merging.
Score: 41

martynaRP commented:
Thank you! It was a really great idea to create such an experience for us :)
